provider/pkg/gen/schema_test.go

```diff
@@ -0,0 +1,189 @@
+	var fileNames []string
+	require.NoError(t, filepath.Walk(filepath.Join(root, "discovery"), func(path string, info os.FileInfo,
+		err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			fileNames = append(fileNames, path)
+		}
+		return nil
+	}))
+
+	metadata, err := loadMetadata()
+	require.NoError(t, err)
+
+	pkgSpec, err := loadSchema()
+	require.NoError(t, err)
+
+	for _, fileName := range fileNames {
+		document, err := readDiscoveryDocument(fileName)
+		require.NoError(t, err)
+
+		if document.Name == "" {
+			continue
+		}
+		module := fmt.Sprintf("%s/%s", document.Name, document.Version)
+		res, ops, err := findResources(fileName, document.Resources, document.Schemas)
+		require.NoError(t, err)
+
+		resourcesForModule := codegen.NewStringSet()
+		for tok := range metadata.Resources {
+			if strings.Contains(tok, module) {
+				resourcesForModule.Add(tok)
+			}
+		}
+
+		for _, tok := range resourcesForModule.SortedValues() {
+			mdResource := metadata.Resources[tok]
+			resource := pkgSpec.Resources[tok]
+			validateOpTemplate := func(resourceOperationType string, op *resources.Operations,
+				ddResource *discoveryDocumentResource) {
+				if op == nil || op.Template == "" {
+					return
+				}
+				tmpl, err := uritemplates.Parse(op.Template)
+				require.NoError(t, err)
+				paramNames := codegen.NewStringSet(tmpl.Names()...)
+
+				var referencedOp *operation
+				switch resourceOperationType {
+				case "CREATE":
+					if ddResource.createMethod.Response.Ref != "" {
+						if createOp, ok := ops[ddResource.createMethod.Response.Ref]; ok {
+							referencedOp = createOp
+						}
+					}
+				case "UPDATE":
+					if ddResource.updateMethod.Response.Ref != "" {
+						if updateOp, ok := ops[ddResource.updateMethod.Response.Ref]; ok {
+							referencedOp = updateOp
+						}
+					}
+				case "DELETE":
+					if ddResource.deleteMethod.Response.Ref != "" {
+						if deleteOp, ok := ops[ddResource.deleteMethod.Response.Ref]; ok {
+							referencedOp = deleteOp
+						}
+					}
+				default:
+					assert.Failf(t, "unexpected resourceOperationType: %q", resourceOperationType)
+				}
+
+				for _, v := range op.Values {
+					if v.Kind == "query" {
+						continue
+					}
+					assert.Containsf(t, paramNames, v.Name, "[%s] [%s] %s.%s not found", resourceOperationType,
+						op.Template, tok, v.Name)
+					expectedName := v.Name
+					if v.SdkName != "" {
+						expectedName = v.SdkName
+					}
+
+					found := false
+					if referencedOp != nil {
+						for name := range referencedOp.schema.Properties {
+							if name == expectedName {
+								found = true
+								break
+							}
+						}
+					}
+
+					for name := range resource.InputProperties {
+						if name == expectedName {
+							found = true
+							break
+						}
+					}
+
+					if found {
+						continue
+					}
+
+					for name := range resource.Properties {
+						if name == expectedName {
+							found = true
+							break
+						}
+					}
```

Comment:
Is the `if found` skip supposed to appear before the `resource.Properties` check?

Reply:
Yeah. I have added a few more comments to the test to make the intent here clearer. Could you take another look?